src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -117,6 +125,32 @@ protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException
         }
     }
 
+    /**
+     * Get file checksum calculating it or retrieving from cache.
+     */
+    private Checksum fileChecksum(File file) throws IOException {
+        // When callers all request the checksum of a large jar the calls to
+        // forURL will all fall through to this method since the first caller's
+        // calculation may take a while. Hence re-check the cache at the start.
+        String location = file.getCanonicalPath();
+        if (checksumsByPath.containsKey(location)) {
+            return checksumsByPath.get(location);
+        }
+
+        Checksum cs;
+        // Synchronize the calculation to throttle the load on master side in
+        // case many slaves are connecting at the same time.
+        // Previously when a large number of slaves connected at the same time
+        // the master would experience a spike in CPU and probably I/O. By caching
+        // the results and synchronizing the calculation of the results this issue addressed.
+        synchronized (this) {
+            cs = Checksum.forFile(file);
+        }
+        checksumsByPath.put(location, cs);
```

Comment:
Again, in the case of unlikely parallel download of different JARs, we may put wrong checksum here.

* Download A 
* Thread 1. Thread interrupt/scheduling
* Thread 1.  Calculate checksum
* Thread 2. Download B, replace A
* Thread 2. Calculate B checksum
* Thread 2. Put B checksum to the cache
* Thread 2. Thread interrupt/scheduling
* Thread 1. Put A checksum to the cache (but the file is B)